server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,465 @@
+        // this method ensures no matter what, we always exit from this function after clearing the voting config exclusion
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+        decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.IN_PROGRESS, new ActionListener<>() {
+            @Override
+            public void onResponse(DecommissionStatus status) {
+                logger.info("updated the decommission status to [{}]", status.toString());
+                // execute nodes decommissioning
+                decommissionController.removeDecommissionedNodes(
+                    filterNodesWithDecommissionAttribute(clusterService.getClusterApplierService().state(), decommissionAttribute, false),
+                    "nodes-decommissioned",
+                    TimeValue.timeValueSeconds(120L),
+                    new ActionListener<Void>() {
+                        @Override
+                        public void onResponse(Void unused) {
+                            clearVotingConfigExclusionAndUpdateStatus(true, true);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            clearVotingConfigExclusionAndUpdateStatus(false, false);
+                        }
+                    }
+                );
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(
+                    () -> new ParameterizedMessage(
+                        "failed to update decommission status for attribute [{}] to [{}]",
+                        decommissionAttribute.toString(),
+                        DecommissionStatus.IN_PROGRESS
+                    ),
+                    e
+                );
+                // since we are not able to update the status, we will clear the voting config exclusion we have set earlier
+                clearVotingConfigExclusionAndUpdateStatus(false, false);
+            }
+        });
+    }
+
+    private void clearVotingConfigExclusionAndUpdateStatus(boolean decommissionSuccessful, boolean waitForRemoval) {
+        decommissionController.clearVotingConfigExclusion(new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void unused) {
+                logger.info(
+                    "successfully cleared voting config exclusion after completing decommission action, proceeding to update metadata"
+                );
+                DecommissionStatus updateStatusWith = decommissionSuccessful ? DecommissionStatus.SUCCESSFUL : DecommissionStatus.FAILED;
+                decommissionController.updateMetadataWithDecommissionStatus(updateStatusWith, statusUpdateListener());
```

Comment:
Note: this is a hard reset as FAILURE which might override a SUCCESSFUL state to FAILURE